test/integration/scheduler/queue_test.go

```diff
@@ -47,6 +51,128 @@ import (
+			},
+			featureEnabled: false,
+			want:           []string{"p1", "p2"},
+		},
+		{
+			name: "feature enabled, one pod carrying scheduling gates",
+			pods: []*v1.Pod{
+				st.MakePod().Name("p1").SchedulingGates([]string{"foo"}).Container("pause").Obj(),
+				st.MakePod().Name("p2").Container("pause").Obj(),
+			},
+			featureEnabled: true,
+			want:           []string{"p2"},
+		},
+		{
+			name: "feature enabled, two pod carrying scheduling gates, and remove gates of one pod",
+			pods: []*v1.Pod{
+				st.MakePod().Name("p1").SchedulingGates([]string{"foo"}).Container("pause").Obj(),
+				st.MakePod().Name("p2").SchedulingGates([]string{"bar"}).Container("pause").Obj(),
+				st.MakePod().Name("p3").Container("pause").Obj(),
+			},
+			featureEnabled:        true,
+			want:                  []string{"p3"},
+			rmPodsSchedulingGates: []int{1}, // remove gates of 'p2'
+			wantPostGatesRemoval:  []string{"p2"},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.PodSchedulingReadiness, tt.featureEnabled)()
+
+			// Use zero backoff seconds to bypass backoffQ.
+			// It's intended to not start the scheduler's queue, and hence to
+			// not start any flushing logic. We will pop and schedule the Pods manually later.
+			testCtx := testutils.InitTestSchedulerWithOptions(
+				t,
+				testutils.InitTestAPIServer(t, "pod-scheduling-gates", nil),
+				0,
+				scheduler.WithPodInitialBackoffSeconds(0),
+				scheduler.WithPodMaxBackoffSeconds(0),
+			)
+			testutils.SyncInformerFactory(testCtx)
+			defer testutils.CleanupTest(t, testCtx)
+
+			cs, ns, ctx := testCtx.ClientSet, testCtx.NS.Name, testCtx.Ctx
+			for _, p := range tt.pods {
+				p.Namespace = ns
+				if _, err := cs.CoreV1().Pods(ns).Create(ctx, p, metav1.CreateOptions{}); err != nil {
+					t.Fatalf("Failed to create Pod %q: %v", p.Name, err)
+				}
+			}
+
+			// Wait for the pods to be present in the scheduling queue.
+			if err := wait.Poll(time.Millisecond*200, wait.ForeverTestTimeout, func() (bool, error) {
+				pendingPods, _ := testCtx.Scheduler.SchedulingQueue.PendingPods()
+				return len(pendingPods) == len(tt.pods), nil
+			}); err != nil {
+				t.Fatal(err)
+			}
+
+			// Pop the expected pods out. They should be de-queueable.
+			for _, wantPod := range tt.want {
+				podInfo := nextPodOrDie(t, testCtx)
+				if got := podInfo.Pod.Name; got != wantPod {
+					t.Errorf("Want %v to be popped out, but got %v", wantPod, got)
+				}
+			}
+
+			if len(tt.rmPodsSchedulingGates) == 0 {
+				return
+			}
+			// Remove scheduling gates from the pod spec.
+			for _, idx := range tt.rmPodsSchedulingGates {
+				patch := `{"spec": {"schedulingGates": null}}`
+				podName := tt.pods[idx].Name
+				if _, err := cs.CoreV1().Pods(ns).Patch(ctx, podName, types.StrategicMergePatchType, []byte(patch), metav1.PatchOptions{}); err != nil {
```

Comment:
not an expert on these things, but I can see that the strategy is `merge``

https://github.com/kubernetes/kubernetes/blob/7b6293b6b6a5662fc37f440e839cf5da8b96e935/staging/src/k8s.io/api/core/v1/types.go#L3340-L3343

is this patch removing the gates?

Reply:
yes, it's removing all scheduling gates.

Reply:
yeah, for strategic merge patch;
* `foo: null` deletes the whole field
* `foo: [{"name":"bar"}]` adds or overwrites the existing entry with name=bar
* `foo: [{"$patch":"delete", "name":"bar"}]` deletes the existing entry with name=bar